tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
+        root_op->getOperand(2), ValueRange({col_index}));
   }
+  b.create<scf::YieldOp>(loc, ValueRange({}));
+
+  b.setInsertionPointToEnd(parallel_op.getBody());
+  b.create<scf::YieldOp>(loc, ValueRange({}));  // parallel
+
+  // remove the root_op if it has no other users except the memref
+  cleanUnusedLhloOps(parent);
+  return success();
+}
+
+Value createSharedMemory(OpBuilder& b, Location loc, int64_t size,
+                         Type elem_type) {
+  auto workgroupMemoryAddressSpace = gpu::AddressSpaceAttr::get(
+      b.getContext(), gpu::GPUDialect::getWorkgroupAddressSpace());
+  auto bufferType =
+      MemRefType::get(ArrayRef<int64_t>{size}, elem_type,
+                      MemRefLayoutAttrInterface{}, workgroupMemoryAddressSpace);
+  auto alloc = b.create<memref::AllocOp>(loc, bufferType);
+  return alloc.getResult();
+}
 
-  for (int var_wi = 0; var_wi < c_tilesize_w; ++var_wi) {
-    b.setInsertionPointToEnd(&if_inbound_op[var_wi].getThenRegion().front());
-    b.create<scf::YieldOp>(loc, yield_values_if_inbound_then[var_wi]);
-    b.setInsertionPointToEnd(&if_inbound_op[var_wi].getElseRegion().front());
-    b.create<scf::YieldOp>(loc, yield_values_if_inbound_else[var_wi]);
+// BLOCK TILE LOOP IMPL: block_x * block_y threads load block_x * block_y *
+// tile_h elements, each thread reduces tile_h elements from gmem to SHM(block_x
+// * block_y), do reduction in SHM and atomic to gmem var_tile_h = 64;
+// num_threads_col = 32;
+// num_threads_row = 8;
+// var_threads = num_threads_col * num_threads_row;
+// num_blocks_col = ceil(var_cols / num_threads_col);
+// num_blocks_row = ceil(var_rows / (var_tile_h * num_threads_row));
+// gridDim.y var_blocks = num_blocks_col * num_blocks_row; for (m = 0; m <
+// var_blocks; ++m) {
+//   for (n = 0; n < var_threads; ++n) {
+//     block_col_index = m % num_blocks_col;
+//     block_row_index = m / num_blocks_col;
+//     thread_col_index = n % num_threads_col;
+//     thread_row_index = n / num_threads_col;
+//     shm_trans_index = thread_col_index * num_threads_row + thread_row_index;
+//     col_index = block_col_index * num_threads_col + thread_col_index;
+//     is_col_valid = col_index < var_cols;
+//     if (is_col_valid) {
+//       accum = init_value;
+//       for (int l = 0; l < var_tile_h; ++l) {
+//         row_index = (thread_row_index  + block_row_index * num_threads_row) *
+//         var_tile_h + l;
+//         is_row_valid = row_index < var_rows;
+//         if(is_row_valid) {
+//           accum = accum + global[row_index, col_index];
+//         } else {
+//           accum = accum;
+//         }
+//       }
+//     } else {
+//       accum = init_value;
+//     }
+//     shm[shm_trans_index] = accum;
+//     for (int stride = num_threads_row / 2; stride > 0; stride /= 2) {
+//       __syncthreads();
+//       if (thread_row_index < stride) {
+//         sdata[shm_trans_index] += sdata[shm_trans_index + stride];
```

Comment:
Use the same name for `shm` and `sdata`.

Reply:
Done.